The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6449537537/job/17507958508) [(plain)](https://github.com/rust-lang/rust/commit/279ce8635af70abe10ba063c15f80ca64c73914b/checks/17507958508/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=daxpedda
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_dbb94f03-88bb-4071-bf2e-85b177ad1d66
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=once-cell-lock-try-insert
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_dbb94f03-88bb-4071-bf2e-85b177ad1d66
GITHUB_REF=refs/pull/116540/merge
GITHUB_REF_NAME=116540/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=e7ee38eaa9d41d06c39f6f8184a0b7806911ccdb
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_dbb94f03-88bb-4071-bf2e-85b177ad1d66
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_dbb94f03-88bb-4071-bf2e-85b177ad1d66
GITHUB_TRIGGERING_ACTOR=daxpedda
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116540/merge
GITHUB_WORKFLOW_SHA=e7ee38eaa9d41d06c39f6f8184a0b7806911ccdb
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
   Compiling cc v1.0.79
   Compiling core v0.0.0 (/checkout/library/core)
   Compiling libc v0.2.148
   Compiling memchr v2.5.0
error: expected one of `:`, `@`, or `|`, found `,`
    |
    |
119 |     pub fn try_insert(&self, value, T) -> Result<&T, (&T, T)> {
    |                                   ^ expected one of `:`, `@`, or `|`
    = note: anonymous parameters are removed in the 2018 edition (see RFC 1685)
help: if this is a parameter name, give it a type
    |
    |
119 |     pub fn try_insert(&self, value: TypeName, T) -> Result<&T, (&T, T)> {
help: if this is a type, explicitly ignore the parameter name
    |
    |
119 |     pub fn try_insert(&self, _: value, T) -> Result<&T, (&T, T)> {


error: expected one of `:`, `@`, or `|`, found `)`
    |
    |
119 |     pub fn try_insert(&self, value, T) -> Result<&T, (&T, T)> {
    |                                      ^ expected one of `:`, `@`, or `|`
    = note: anonymous parameters are removed in the 2018 edition (see RFC 1685)
help: if this is a parameter name, give it a type
    |
    |
119 |     pub fn try_insert(&self, value, T: TypeName) -> Result<&T, (&T, T)> {
help: if this is a type, explicitly ignore the parameter name
    |
    |
119 |     pub fn try_insert(&self, value, _: T) -> Result<&T, (&T, T)> {

   Compiling compiler_builtins v0.1.101
   Compiling unwind v0.0.0 (/checkout/library/unwind)
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling std v0.0.0 (/checkout/library/std)
error[E0711]: feature `once_cell` is declared stable since 1.73.0-beta.1, but was previously declared stable since 1.70.0
    |
    |
118 |     #[stable(feature = "once_cell", since = "CURRENT_RUSTC_VERSION")]

For more information about this error, try `rustc --explain E0711`.
error: could not compile `core` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

--- a/library/core/src/cell/once.rs
+++ b/library/core/src/cell/once.rs
@@ -87,10 +87,40 @@ impl<T> OnceCell<T> {
     #[inline]
     #[stable(feature = "once_cell", since = "1.70.0")]
     pub fn set(&self, value: T) -> Result<(), T> {
-        // SAFETY: Safe because we cannot have overlapping mutable borrows
-        let slot = unsafe { &*self.inner.get() };
-        if slot.is_some() {
-            return Err(value);
+        match self.try_insert(value) {
+            Ok(_) => Ok(()),
+            Err((_, value)) => Err(value),
+        }
+    }
+
+    /// Sets the contents of the cell to `value` if the cell was empty, then
+    /// returns a reference to it.
+    ///
+    /// # Errors
+    ///
+    /// This method returns `Ok(&value)` if the cell was empty and
+    /// `Err(&current_value, value)` if it was full.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(once_cell_try_insert)]
+    ///
+    /// use std::cell::OnceCell;
+    ///
+    /// let cell = OnceCell::new();
+    /// assert!(cell.get().is_none());
+    ///
+    /// assert_eq!(cell.try_insert(92), Ok(&92));
+    /// assert_eq!(cell.try_insert(62), Err((&92, 62)));
+    ///
+    /// assert!(cell.get().is_some());
+    /// ```
+    #[inline]
+    #[unstable(feature = "once_cell_try_insert", issue = "116693")]
+    pub fn try_insert(&self, value: T) -> Result<&T, (&T, T)> {
+        if let Some(old) = self.get() {
+            return Err((old, value));
         }
 
         // SAFETY: This is the only place where we set the slot, no races
@@ -99,7 +129,7 @@ impl<T> OnceCell<T> {
         // maintains the `inner`'s invariant.
         let slot = unsafe { &mut *self.inner.get() };
         *slot = Some(value);
-        Ok(())
+        Ok(self.get().unwrap())
     }
 
     /// Gets the contents of the cell, initializing it with `f`

--- a/library/core/src/cell/once.rs
+++ b/library/core/src/cell/once.rs
@@ -128,8 +128,7 @@ impl<T> OnceCell<T> {
         // checked that slot is currently `None`, so this write
         // maintains the `inner`'s invariant.
         let slot = unsafe { &mut *self.inner.get() };
-        *slot = Some(value);
-        Ok(self.get().unwrap())
+        Ok(slot.insert(value))
     }
 
     /// Gets the contents of the cell, initializing it with `f`
@@ -213,10 +212,9 @@ impl<T> OnceCell<T> {
         let val = outlined_call(f)?;
         // Note that *some* forms of reentrant initialization might lead to
         // UB (see `reentrant_init` test). I believe that just removing this
-        // `assert`, while keeping `set/get` would be sound, but it seems
+        // `panic`, while keeping `try_insert` would be sound, but it seems
         // better to panic, rather than to silently use an old value.
-        assert!(self.set(val).is_ok(), "reentrant init");
-        Ok(self.get().unwrap())
+        if let Ok(val) = self.try_insert(val) { Ok(val) } else { panic!("reentrant init") }
     }
 
     /// Consumes the cell, returning the wrapped value.

--- a/library/std/src/sync/once_lock.rs
+++ b/library/std/src/sync/once_lock.rs
@@ -126,11 +126,48 @@ impl<T> OnceLock<T> {
     #[inline]
     #[stable(feature = "once_cell", since = "1.70.0")]
     pub fn set(&self, value: T) -> Result<(), T> {
+        match self.try_insert(value) {
+            Ok(_) => Ok(()),
+            Err((_, value)) => Err(value),
+        }
+    }
+
+    /// Sets the contents of this cell to `value` if the cell was empty, then
+    /// returns a reference to it.
+    ///
+    /// May block if another thread is currently attempting to initialize the cell. The cell is
+    /// guaranteed to contain a value when set returns, though not necessarily the one provided.
+    ///
+    /// Returns `Ok(&value)` if the cell was empty and `Err(&current_value, value)` if it was full.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(once_cell_try_insert)]
+    ///
+    /// use std::sync::OnceLock;
+    ///
+    /// static CELL: OnceLock<i32> = OnceLock::new();
+    ///
+    /// fn main() {
+    ///     assert!(CELL.get().is_none());
+    ///
+    ///     std::thread::spawn(|| {
+    ///         assert_eq!(CELL.try_insert(92), Ok(&92));
+    ///     }).join().unwrap();
+    ///
+    ///     assert_eq!(CELL.try_insert(62), Err((&92, 62)));
+    ///     assert_eq!(CELL.get(), Some(&92));
+    /// }
+    /// ```
+    #[inline]
+    #[unstable(feature = "once_cell_try_insert", issue = "116693")]
+    pub fn try_insert(&self, value: T) -> Result<&T, (&T, T)> {
         let mut value = Some(value);
-        self.get_or_init(|| value.take().unwrap());
+        let res = self.get_or_init(|| value.take().unwrap());
         match value {
-            None => Ok(()),
-            Some(value) => Err(value),
+            None => Ok(res),
+            Some(value) => Err((res, value)),
         }
     }
 

